Feat: merge ODK central work

jniles opened the pull request:
This PR adds in [ODK Central](https://docs.getodk.org/central-install/) functionality to the BHIMA server.  Here is what the ODK administration interface looks like:

![image](https://user-images.githubusercontent.com/896472/150951277-34cf068b-9eca-4363-a485-ad135ad18df1.png)

You'll need to set your own username/password and ODK Central URL to test it out.

This installation is very bare-bones.  The way this works is somewhat documented in https://github.com/IMA-WorldHealth/bhima/issues/6235#issuecomment-1014303259.

This PR is plucked from work done by @jniles and @mbayopanda in `sprint-prosani`.

Closes #6235.
Closes #6302.
Closes #6322.

---

**TESTING**  (based on session with @jniles )
- Use bhima test (rebuild db, etc)
- Go to Administration > ODK Settings
  - Enter the ODK Central login credentials in the left-most panel (ODK Integration Settings)
  - In the middle panel, click on
    - Synchronize enterprise settings
    - Synchronize users
    - Synchronize forms
- Go to Adminstration > User Management
   - In the action menu for Superuser, click on the [ODK QR Code] link to display the QR code (leave it up on your screen)
- On your mobile phone
  - If you have not installed ODK Collect, do that
  - In ODK Collect, click on the [Configure with QR code] button on the entry screen. This should bring up the "Test Enterprise" page
  - Click on [Get Blank Form].  It should show that it has downloaded a form
  - On the bottom right of the page, click on the [Get Selected] button.  This downloads the form and sends you back to the "Test Enterprise" page.
  - Click on [Fill Blank Form] and select the form.  This should start filling out the form.
  - Fill out the form
     - For Nombre des articles, enter 1
     - For the barcode, first go to Bhima (Stock Lots) and display the barcode for lot VITAMIN-A.  Then scan that barcode with your phone.  When you click next, it should show the lot you selected.
     - Finish the form ([X] Mark form as finalize, press [Save Form and Exit])
  - Back on the "Test Enterprise" page, click [Send Finalized Form].  Select the form you just completed and press [Send Selected] on the bottom right of the page.
  - Back on Bhima go to Administration > ODK Settings
  - Click on:  Synchronize Submissions [Sync]
  - Go to Stock > Stock Movements and verify that you have a new stock entry in Depot Tertiaire

jniles committed:
feat(odk): implement ODK Central Link

This commit implements the ODK Central link with BHIMA, developped
during the `prosani-sprint`.  The goal is to have a functional, optional
link to ODK Central that can be merged into the BHIMA repository.

Closes #6235.

jniles committed:
fix(users): skip users without email addresses

This commit skips users in the database without email addresses.

Closes #6322.

jniles committed:
chore(odk): disable fields on loading

This commit disables the sync buttons when synchronisation is not
finished yet.

jniles committed:
chore(odk): add better loading/default text

This commit adds better loading and default text to the ODK Central
configuration page.

jniles committed:
chore: final touches

This commit cleans up the ODK Central work, having moved some items out
of the odk-central file to be put in the ODK Central API library.

jmcameron reviewed (commented):
Could not sync with the server.  Tried logging manually and that worked.  But when I carefully enter the ODK central server credentials and try "Sync Enterprise", I get 403 errors.

jniles committed:
docs: add documentation on ODK Central

jniles committed:
chore: remove old code

This commit cleans up old code and gets the ODK portion totally ready
for submission.

jmcameron commented on server/models/bhima.sql:
```diff
@@ -178,7 +178,8 @@ INSERT INTO unit VALUES
   (304, '[SETTINGS] Settings', 'TREE.PAYROLL_SETTINGS', 'Payroll Settings', 57, '/payroll/setting'),
-  (305, 'Avg Medical Costs Per Patient', 'TREE.AVERAGE_MED_COST_REPORT', 'Report of avg med costs', 282, '/reports/avg_med_costs_per_patient');
+  (305, 'Avg Medical Costs Per Patient', 'TREE.AVERAGE_MED_COST_REPORT', 'Report of avg med costs', 282, '/reports/avg_med_costs_per_patient'),
+  (306, '[SETTINGS] ODK Settings', 'TREE.ODK_SETTINGS', 'ODK Settings', 1, '/admin/odk-settings');
```
While using the [Settings] option to add the gear icon for the ODK Settings looks nice, it a bit odd in the context of the other Administration menu items.  Several of the items in that menu are for settings (eg Enterprise) and are not marked with the gear icon.  So the ODK Settings is not consistent with the rest.  So I suggest either removing [SETTINGS] or adding to other items in the Administration menu that are for settings.  Also, in reality, the ODK Settings item is about more than settings.  Only the left most panel is for settings.  The rest are for actions, so this page is not ALL settings.

jniles replied:
Yeah, this is a good point.  Let me do that.

jmcameron commented on server/models/migrations/next/migrate.sql:
```diff
@@ -1,11 +1,35 @@
- * @description: update migrate.sql file for
- * @date: 2022-01-24
-*/
-UPDATE report SET `report_key` = 'analysis_auxiliary_cashbox', `title_key` = 'REPORT.ANALYSIS_AUX_CASHBOX.TITLE'
```
It was not obvious why these got deleted.

jniles replied:
Probably a failed merge.  Let me fix it.

jmcameron reviewed (commented):
I found the term "ODK Central" a little vague.  As I was trying to figure out ODK, it was not obvious if this was part of BHIMA or not.  So I suggest renaming this to "BHIMA ODK Server" or something similar.

Obviously this is a bare-bones ODK server.   However (in another PR), we need a registry for ODK forms, submissions, aggregated data, etc.

I did a basic review of the code.  See the inline code comments.

I also tested this (per the procedure outlined in the PR description) with assistance from @jniles.  Other than an online server database issue that @jniles fixed behind the scenes, it all worked nicely.

We found that the QR code for the ODK server settings works well.  However, the BHIMA lot barcodes seem to be a little small for reliable scanning on my cell phone.  So I suggest increasing the lot barcode display.  If that is not done in this PR, I suggest creating a new issue for it.

In practice, I suspect the layout of this page will evolve.  It seems like it should perhaps be a "ODK Campaign" registry and the Settings and sync actions might fall under the main Menu button?

jniles committed:
chore: refactor from review

This commit refactors code from the review.

jniles committed:
fix(odk): auto-assign form to user(s) on create

This commit automatically assigns teh forms to the users on creation of
the form.

jniles committed:
chore: bump deps
